### PR TITLE
Background add-in and ribbon APIs

### PR DESCRIPTION
**Pull request by sumurthy**
We are releasing background addins API and Office ribbon APIs. This definition adds the API details. 

Please fill in this template.
- [x ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ x] Test the change in your own code. (Compile and run.)
- [x ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.microsoft.com/en-us/office/dev/add-ins/overview/office-add-ins
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/office-js-preview/index.d.ts [hunk 1]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+         * @param input Represents the updates to be made to the ribbon. Note that only the changes specified in the input parameter are made.
+         */
+        requestUpdate(input: RibbonUpdaterData): Promise<void>;
+    }
+    /**
+     * Specifies changes to the ribbon, such as the enabled or disabled status of a button.
+     */
+    interface RibbonUpdaterData {
+        /**
+         * Collection of tabs whose state is set with the call of requestUpdate.
```
Comment (Rick-Kirkham) [inline types/office-js-preview/index.d.ts]:
```suggestion
         * Collection of tabs whose state is set with the call of `requestUpdate`.
```

**Review by Rick-Kirkham — CHANGES_REQUESTED** [review]

File: types/office-js-preview/index.d.ts [hunk 2]
```diff
@@ -325,7 +325,47 @@ declare namespace Office {
...
+    const ribbon: Ribbon;
+    /**
+     * Check if the specified requirement set is supported by the host Office application.
+     * @param name - Set name; e.g., "MatrixBindings".
+     * @param minVersion - The minimum required version; e.g., "1.4".
+     */
+    function isSetSupported(name: string, minVersion?: string): boolean;
     // Enumerations
+    /**
+     * Startup behavior of the add-in.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Can we rephrase this to describe what's happening rather than repeating the enum name? Like, this is the behavior of the add-in when the Office host starts or when the add-in is opened?

File: types/office-js-preview/index.d.ts [hunk 3]
```diff
@@ -325,7 +325,47 @@ declare namespace Office {
...
+    /**
+     * Represents the add-in.
+     */
+    const addin: Addin;
+    /**
+     * Represents the ribbon associated with the Office application.
+     */
+    const ribbon: Ribbon;
+    /**
+     * Check if the specified requirement set is supported by the host Office application.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Grammar consistency: "Check" -> "Checks"

File: types/office-js-preview/index.d.ts [hunk 4]
```diff
@@ -325,7 +325,47 @@ declare namespace Office {
...
+     * @param minVersion - The minimum required version; e.g., "1.4".
+     */
+    function isSetSupported(name: string, minVersion?: string): boolean;
     // Enumerations
+    /**
+     * Startup behavior of the add-in.
+     */
+    enum StartupBehavior {
+        /**
+         * Nothing.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Can we be more descriptive here? Maybe "The add-in does not load until opened by the user."

File: types/office-js-preview/index.d.ts [hunk 5]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+     * Message used in the visibility mode changed event.
+     */
+    interface VisibilityModeChangedMessage {
+        /**
+         * Visibility changed state.
+         */
+        visibilityMode: Office.VisibilityMode;
+    }
+    /**
+     * Function type to turn off the event.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: "the event" - which event is that?

File: types/office-js-preview/index.d.ts [hunk 6]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
         */
         value: T;
     }
+    /**
+     * Message used in the visibility mode changed event.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Later, you use "VisbilityModeChanged event" to describe this. I like using the actual name. Can we surround it with code ticks (`)?

File: types/office-js-preview/index.d.ts [hunk 7]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+         * @returns A promise that is resolved when the UI is shown.
+         */
+        showAsTaskpane(): Promise<void>;
+        /**
+         * Hides the task pane.
+         * @returns A promise that is resolved when the UI is hidden.
+         */
+        hide(): Promise<void>;
+        /**
+         * Adds a listener for the VisbilityModeChanged event.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Can we surround VisibilityModeChanged with code ticks (`)?

File: types/office-js-preview/index.d.ts [hunk 8]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+        /**
+         * Shows the task pane associated with the add-in.
+         * @returns A promise that is resolved when the UI is shown.
+         */
+        showAsTaskpane(): Promise<void>;
+        /**
+         * Hides the task pane.
+         * @returns A promise that is resolved when the UI is hidden.
+         */
+        hide(): Promise<void>;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Shouldn't this method be `hideTaskpane`?

File: types/office-js-preview/index.d.ts [hunk 9]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+        setStartupBehavior(behavior: Office.StartupBehavior): Promise<void>;
+        /**
+         * Get the current startup behavior for the add-in.
+         */
+        getStartupBehavior(): Promise<Office.StartupBehavior>;
+        /**
+         * Shows the task pane associated with the add-in.
+         * @returns A promise that is resolved when the UI is shown.
+         */
+        showAsTaskpane(): Promise<void>;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Why is this `showAsTaskpane` and not `showTaskpane`?

File: types/office-js-preview/index.d.ts [hunk 10]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+     * Function type to turn off the event.
+     */
+    type RemoveEventListener = () => Promise<void>;
+    /**
+     * Represents add-in level functionality for operating or configuring various aspects of the add-in.
+     */
+    interface Addin {
+        /**
+         * Set the startup behavior for the add-in for when the document is opened next time.
+         * @param behavior Specifies startup behavior of the add-in.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: TSDoc prefers a hyphen between the parameter name and the description.
https://github.com/Microsoft/tsdoc

File: types/office-js-preview/index.d.ts [hunk 11]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+         * Adds a listener for the VisbilityModeChanged event.
+         * @param listener The listener function that is called when the event is emitted. This function takes in a message for the receiving component.
+         * @returns A promise that resolves when the listener is added.
+         */
+        onVisibilityModeChanged(
+            listener: (message: VisibilityModeChangedMessage) => void,
+        ): Promise<RemoveEventListener>;
+    }
+    /**
+     * Ribbon interface contains all the functionality provided to manage the state of the OFfice ribbon.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Grammar nit: It sounds a little awkward to say "Ribbon interface contains". How about "An interface that contains"?
That way, we can avoid having to worry about denoting the interface name with code ticks too.

File: types/office-js-preview/index.d.ts [hunk 12]
```diff
@@ -476,6 +516,114 @@ declare namespace Office {
...
+     */
+    interface Ribbon {
+        /**
+         * Sends a request to Office to update the ribbon.
+         * Note that this API is only to request for an update. The actual UI update to the ribbon is controlled by the Office application and hence the exact timing of the ribbon update (or refresh) cannot be determined by the completion of this API.
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Typo: "request for an update" -> "request an update"

**Review by AlexJerabek — CHANGES_REQUESTED** [review]

File: types/office-js-preview/index.d.ts (hunk 8 above)
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: The UI in the future could be shown as dialog or taskpane. for now only taskpane is supported. So, we'd like to have a generic hide API.

File: types/office-js-preview/index.d.ts (hunk 5 above)
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: It's a generic type and can be used across any events that returns this function.
this is to avoid having to need the object to call the event off method.

Comment (AlexJerabek) [reply types/office-js-preview/index.d.ts]: Then shouldn't it be "an event" instead of "the event"?

**Review by Rick-Kirkham — APPROVED** [review]

File: types/office-js-preview/index.d.ts [hunk 13]
```diff
@@ -325,7 +325,47 @@ declare namespace Office {
      * @param useShortNamespace True to use the shortcut alias; otherwise false to disable it. The default is true.
      */
     function useShortNamespace(useShortNamespace: boolean): void;
+    /**
+     * Represents the add-in.
```
[the remark this answers is not in the document]
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: It is in a way add-ins runtime settings. We wanted to avoid using the work runtime. 
Using this object you can control addins display, startup behavior, etc.

[the remark this answers is not in the document]
Comment (AlexJerabek) [reply types/office-js-preview/index.d.ts]: Can we add that sort of information to the description?

File: types/office-js-preview/index.d.ts (hunk 2 above)
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: done

File: types/office-js-preview/index.d.ts (hunk 3 above)
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: Done

File: types/office-js-preview/index.d.ts (hunk 6 above)
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: Done

File: types/office-js-preview/index.d.ts (hunk 7 above)
Comment (sumurthy) [reply types/office-js-preview/index.d.ts]: Done

File: types/office-js-preview/index.d.ts [hunk 14]
```diff
@@ -358,11 +358,11 @@ declare namespace Office {
      */
     enum VisibilityMode {
         /**
-         * Hidden
+         * UI is Hidden
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: "UI is Hidden" -> "UI is hidden."

File: types/office-js-preview/index.d.ts [hunk 15]
```diff
@@ -358,11 +358,11 @@ declare namespace Office {
...
          */
         hidden = 'Hidden',
         /**
-         * Visible
+         * Displayed as taskpane
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Please add a period to the end of this description.

**Review by AlexJerabek — APPROVED** [review]
Approving! Thanks for making those changes. There's a couple minor things that do not need a re-review.

**Comment by Rick-Kirkham** [comment]
PLEASE WAIT FOR SECOND REVIEWER APPROVAL BEFORE MERGING.

**Comment by AlexJerabek** [comment]
@Rick-Kirkham, in addition to my review?

**Comment by Rick-Kirkham** [comment]
> @Rick-Kirkham, in addition to my review?

My bad. I didn't realize that you'd already approved. I deleted my comment.
